rpc.go

```diff
@@ -158,6 +159,182 @@ func (c *client) getRegionAndClientForRPC(ctx context.Context, rpc hrpc.Call) (
 	}
 }
 
+var (
+	// NotExecutedError is returned when an RPC in a batch is not
+	// executed due to encountering a different error in the batch.
+	NotExecutedError = errors.New(
+		"RPC in batch not executed due to another error")
+)
+
+// SendBatch will execute all the Calls in batch. Every Call must have
+// the same table and must be Batchable.
+//
+// SendBatch will discover the correct region and region server for
+// each Call and dispatch the Calls accordingly. SendBatch is not an
+// atomic operation. Some calls may fail and others succeed. Calls
+// sharing a region will execute in the order passed into SendBatch.
+//
+// SendBatch returns a slice of [hrpc.RPCResult] each containing a
+// response and an error. The results will be returned in the same
+// order as the Calls in the batch, in other words the i'th result
+// will be for the i'th call. A nil error means the Call executed
+// successfully. allOK is true if all calls completed successfully,
+// and false if any calls failed and the errors in the results need to
+// be checked.
+func (c *client) SendBatch(ctx context.Context, batch []hrpc.Call) (
+	res []hrpc.RPCResult, allOK bool) {
+	if len(batch) == 0 {
+		return nil, true
+	}
+
+	allOK = true
+
+	start := time.Now()
+	description := "SendBatch"
+	ctx, sp := observability.StartSpan(ctx, description)
+	defer func() {
+		result := "ok"
+		if !allOK {
+			result = "error"
+			sp.SetStatus(codes.Error, "batch error")
+		}
+
+		o := operationDurationSeconds.WithLabelValues(description, result)
+
+		observability.ObserveWithTrace(ctx, o, time.Since(start).Seconds())
+		sp.End()
+	}()
+
+	table := batch[0].Table()
+	res = make([]hrpc.RPCResult, len(batch))
+	rpcToRes := make(map[hrpc.Call]int, len(batch))
+	for i, rpc := range batch {
+		// map Call to index in res so that we can set the correct
+		// result as Calls complete
+		if j, dup := rpcToRes[rpc]; dup {
+			res[i].Error = fmt.Errorf("duplicate call in batch at index %d", j)
+			allOK = false
+			continue
+		}
+		rpcToRes[rpc] = i
+
+		// Initialize res with NotExecutedError. As RPCs are executed this
+		// will be replaced by a more specific error or nil if no error
+		// occurs.
+		res[i].Error = NotExecutedError
+
+		if !bytes.Equal(rpc.Table(), table) {
+			res[i].Error = fmt.Errorf("multiple tables in batch request: %q and %q",
+				string(table), string(rpc.Table()))
+			allOK = false
+		} else if b, batchable := rpc.(hrpc.Batchable); !batchable || b.SkipBatch() {
+			res[i].Error = errors.New("non-batchable call passed to SendBatch")
+			allOK = false
+		}
+	}
+	if !allOK {
+		return res, allOK
+	}
+
+	rpcByClient, ok := c.findClients(ctx, batch, res)
```

Comment:
Could interesting to add an histogram to track how many regions are "targeted" by batch of RPC. To be added in following PR maybe.